docs/source/minimal_cpp_demo/meson.build

```diff
@@ -0,0 +1,25 @@
+project(
+    'minimal_cpp_demo',
+    'cpp',
+    version: '0.1',
+    default_options: ['cpp_std=c++14']
+)
+
+pymod = import('python')
+py3_inst = pymod.find_installation('python3')
+
+pybind11_config = find_program('pybind11-config')
+pybind11_config_ret = run_command(pybind11_config, ['--includes'], check: true)
+pybind11 = declare_dependency(
+    include_directories: [pybind11_config_ret.stdout().split('-I')[-1].strip()],
+)
```

Comment:
Thank you very much @eli-schwartz, this helps a lot.

If I may ask, it seems that `meson` finds `pybind11` but not `pybind11-config1`:

```
pybind11-config found: NO
Found CMake: /usr/bin/cmake (3.22.1)
Run-time dependency pybind11 found: YES 2.9.2
```

Reply:
Yes, pybind11 upstream installs 3 different methods:
- a python script called `pybind11-config`
- a pkg-config file called `pybind11.pc`
- a cmake file called `pybind11Config.cmake`

They all report the same information, and meson will try all 3 to make sure it can find the installed pybind11 headers. It will look for the python script before cmake, though.

It should not matter *how* you find pybind11, only that you do find it (possibly with a certain minimum version). Although if you want to specifically require a given method, you can do this:
```meson
dependency('pybind11', method: 'config-tool')
```

I see no real reason to deny the use of a cmake-detected copy of the pybind11 headers, though.

Reply:
I notice you added pybind11-global to the dependencies. That's what provides the cmake config, and it's not strictly necessary as long as the pybind11-config script is installed to $PATH, which is why meson prefers that script if it can find it.

I'm not sure how `pip` build isolation plays with scripts installed to PATH -- I think some old versions of pip may have been buggy here?

Reply:
Great, thank you.

I guess we could add a note about using a recent version of `pip`, or suggest another way of installing pybind11.

Reply:
You can create a "subprojects/" directory, and run `meson wrap install pybind11`. It installs the following file:

```ini
# $ cat subprojects/pybind11.wrap
[wrap-file]
directory = pybind11-2.10.4
source_url = https://github.com/pybind/pybind11/archive/refs/tags/v2.10.4.tar.gz
source_filename = pybind11-2.10.4.tar.gz
source_hash = 832e2f309c57da9c1e6d4542dedd34b24e4192ecb4d62f6f4866a737454c9970
patch_filename = pybind11_2.10.4-1_patch.zip
patch_url = https://wrapdb.mesonbuild.com/v2/pybind11_2.10.4-1/get_patch
patch_hash = 9489d0cdc1244078a3108c52b4591a6f07f3dc30ca7299d3a3c42b84fa763396
source_fallback_url = https://github.com/mesonbuild/wrapdb/releases/download/pybind11_2.10.4-1/pybind11-2.10.4.tar.gz
wrapdb_version = 2.10.4-1

[provide]
pybind11 = pybind11_dep
```

If this file is present, then meson will know how to "provide" pybind11 by downloading this tarball, securely checksumming it, and extracting it to get a meson subproject that "builds" pybind11 and makes it available. (Of course, it is header-only, so this subproject just assigns the header directory and returns the result).

See https://mesonbuild.com/Wrap-dependency-system-manual.html for more discussion on how this works.

tl;dr -- vendoring your dependencies is a common approach to "things should just work out of the box". Meson implements a kind of vendoring that:
- supports both getting an existing installed copy, and using the vendored copy
- doesn't require modifying build files to add vendored copies
- doesn't require checking lots of third-party source code into git
- provides automatic fine-grained control in case you would like to choose whether to use system or vendored copies